File: ChemicalDispersionModeling.Desktop/Services/DispersionModelingService.cs
using ChemicalDispersionModeling.Core.Models;
using ChemicalDispersionModeling.Core.Services;
using Microsoft.Extensions.Logging;

namespace ChemicalDispersionModeling.Desktop.Services;

/// <summary>
/// Implementation of atmospheric dispersion modeling using Gaussian plume model
/// </summary>
public class DispersionModelingService : IDispersionModelingService
{
    private readonly ILogger<DispersionModelingService> _logger;

    public DispersionModelingService(ILogger<DispersionModelingService> logger)
    {
        _logger = logger;
    }

    public async Task<IEnumerable<DispersionResult>> CalculateGaussianPlumeAsync(Release release, IEnumerable<Receptor> receptors, WeatherData weather)
    {
        return await Task.Run(() =>
        {
            try
            {
                _logger.LogInformation($"Calculating Gaussian plume for {receptors.Count()} receptors");
                
                var results = new List<DispersionResult>();
                var effectiveReleaseRate = CalculateEffectiveReleaseRate(release, release.Chemical!, weather);
                var plumeRise = CalculatePlumeRise(release, release.Chemical!, weather);
                var effectiveHeight = release.ReleaseHeight + plumeRise;
                var stabilityClass = DetermineStabilityClass(weather);
                
                foreach (var receptor in receptors)
                {
                    var distance = CalculateDistance(release.Latitude, release.Longitude, receptor.Latitude, receptor.Longitude);
                    
                    if (distance < 1.0) distance = 1.0; // Minimum distance to avoid singularity
                    
                    var concentration = CalculateGaussianConcentration(
                        effectiveReleaseRate,
                        distance,
                        effectiveHeight,
                        weather.WindSpeed,
                        stabilityClass);
                    
                    var result = new DispersionResult
                    {
                        Latitude = receptor.Latitude,
                        Longitude = receptor.Longitude,
                        Concentration = concentration,
                        ConcentrationUnit = "mg/m³",
                        CalculationTime = DateTime.Now,
                        WindSpeed = weather.WindSpeed,
                        WindDirection = weather.WindDirection,
                        Temperature = weather.Temperature,
                        RiskLevel = DetermineRiskLevel(concentration, release.Chemical!),
                        ModelUsed = "Gaussian Plume",
                        ReleaseId = release.Id,
                        ReceptorId = receptor.Id
                    };
                    
                    results.Add(result);
                }
                
                _logger.LogInformation($"Calculated concentrations for {results.Count} receptors");
                return results;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error calculating Gaussian plume");
                throw;
            }
        });
    }

    public async Task<IEnumerable<DispersionResult>> CalculateDispersionGridAsync(Release release, WeatherData weather, double gridSize = 100, double maxDistance = 10000)
    {
        return await Task.Run(() =>
        {
            try
            {
                _logger.LogInformation($"Calculating dispersion grid with {gridSize}m resolution, max distance {maxDistance}m");
                
                var results = new List<DispersionResult>();
                var effectiveReleaseRate = CalculateEffectiveReleaseRate(release, release.Chemical!, weather);
                var plumeRise = CalculatePlumeRise(release, release.Chemical!, weather);
                var effectiveHeight = release.ReleaseHeight + plumeRise;
                var stabilityClass = DetermineStabilityClass(weather);
                
                // Convert wind direction to radians for calculations
                var windDirectionRad = weather.WindDirection * Math.PI / 180.0;
                
                // Clear existing results and create wind-oriented plume
                results.Clear();
                
                // Generate plume shape based on wind direction
                for (double downwindDistance = 10; downwindDistance <= maxDistance; downwindDistance += gridSize)
                {
                    // Calculate crosswind standard deviation (sigma_y) based on distance and stability
                    var sigmaY = CalculateCrosswindDispersion(downwindDistance, stabilityClass);
                    var sigmaZ = CalculateVerticalDispersion(downwindDistance, stabilityClass);
                    
                    // Generate crosswind points for this downwind distance
                    var crosswindSpan = Math.Min(sigmaY * 6, maxDistance); // 6 sigma covers ~99.7% of plume
                    var crosswindPoints = (int)(crosswindSpan / gridSize);
                    
                    for (int j = -crosswindPoints; j <= crosswindPoints; j++)
                    {
                        var crosswindOffset = j * gridSize;
                        
                        // Transform coordinates based on wind direction
                        var offsetX = downwindDistance * Math.Sin(windDirectionRad) + crosswindOffset * Math.Cos(windDirectionRad);
                        var offsetY = downwindDistance * Math.Cos(windDirectionRad) - crosswindOffset * Math.Sin(windDirectionRad);
                        
                        // Calculate actual lat/lon from offset
                        var lat = release.Latitude + (offsetY / 111000.0);
                        var lon = release.Longitude + (offsetX / (111000.0 * Math.Cos(release.Latitude * Math.PI / 180.0)));
                        
                        // Calculate realistic Gaussian concentration with crosswind decay
                        var concentration = CalculateWindOrientedConcentration(
                            effectiveReleaseRate,
                            downwindDistance,
                            crosswindOffset,
                            effectiveHeight,
                            weather.WindSpeed,
                            sigmaY,
                            sigmaZ);
                        
                        if (concentration > 0.001) // Only add points with significant concentration
                        {
                            var result = new DispersionResult
                            {
                                Latitude = lat,
                                Longitude = lon,
                                Concentration = concentration,
                                ConcentrationUnit = "mg/m³",
                                CalculationTime = DateTime.Now,
                                WindSpeed = weather.WindSpeed,
                                WindDirection = weather.WindDirection,
                                Temperature = weather.Temperature,
                                RiskLevel = DetermineRiskLevel(concentration, release.Chemical!),
                                ModelUsed = "Gaussian Plume Grid",
                                ReleaseId = release.Id
                            };
                            
                            results.Add(result);
                        }
                    }
                }
                
                _logger.LogInformation($"Generated {results.Count} grid points with significant concentrations");
                return results;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error calculating dispersion grid");
                throw;
            }
        });
    }

    public async Task<IEnumerable<DispersionResult>> CalculateTimeVaryingDispersionAsync(Release release, IEnumerable<Receptor> receptors, IEnumerable<WeatherData> weatherSequence)
    {
        var allResults = new List<DispersionResult>();
        
        foreach (var weather in weatherSequence)
        {
            var timeResults = await CalculateGaussianPlumeAsync(release, receptors, weather);
            allResults.AddRange(timeResults);
        }
        
        return allResults;
    }

    public double CalculateEffectiveReleaseRate(Release release, Chemical chemical, WeatherData weather)
    {
        // Simple conversion from release rate to effective emission rate
        // This would normally account for chemical properties, temperature, etc.
        var baseRate = release.ReleaseRate ?? 1.0; // kg/s, default to 1 kg/s
        
        // Convert to mg/s for concentration calculations
        return baseRate * 1000000; // mg/s
    }

    public double CalculatePlumeRise(Release release, Chemical chemical, WeatherData weather)
    {
        // Simplified plume rise calculation
        // This would normally use more complex formulas based on buoyancy and momentum
        var releaseTemp = release.InitialTemperature ?? weather.Temperature + 20; // Default to 20°C above ambient
        var deltaT = releaseTemp - weather.Temperature;
        var buoyancyRise = deltaT > 0 ? deltaT * 0.5 : 0; // Very simplified
        
        return Math.Max(0, buoyancyRise);
    }

    public (double SigmaY, double SigmaZ) CalculateDispersionCoefficients(string stabilityClass, double distance, double height)
    {
        // Pasquill-Gifford dispersion coefficients
        var x = distance / 1000.0; // Convert to km for P-G formulas
        
        double sigmaY, sigmaZ;
        
        switch (stabilityClass.ToUpper())
        {
            case "A": // Very unstable
                sigmaY = 213 * x * Math.Pow(1 + 0.0001 * distance, -0.5);
                sigmaZ = 213 * x * Math.Pow(1 + 0.0001 * distance, -0.5);
                break;
            case "B": // Moderately unstable
                sigmaY = 156 * x * Math.Pow(1 + 0.0001 * distance, -0.5);
                sigmaZ = 156 * x * Math.Pow(1 + 0.0001 * distance, -0.5);
                break;
            case "C": // Slightly unstable
                sigmaY = 104 * x * Math.Pow(1 + 0.0001 * distance, -0.5);
                sigmaZ = 104 * x * Math.Pow(1 + 0.0001 * distance, -0.5);
                break;
            case "D": // Neutral
                sigmaY = 68 * x * Math.Pow(1 + 0.0001 * distance, -0.5);
                sigmaZ = 68 * x * Math.Pow(1 + 0.0001 * distance, -0.5);
                break;
            case "E": // Slightly stable
                sigmaY = 50.5 * x * Math.Pow(1 + 0.0001 * distance, -0.5);
                sigmaZ = 50.5 * x * Math.Pow(1 + 0.0001 * distance, -0.5);
                break;
            case "F": // Moderately stable
                sigmaY = 34 * x * Math.Pow(1 + 0.0001 * distance, -0.5);
                sigmaZ = 34 * x * Math.Pow(1 + 0.0001 * distance, -0.5);
                break;
            default:
                sigmaY = 68 * x * Math.Pow(1 + 0.0001 * distance, -0.5); // Default to neutral
                sigmaZ = 68 * x * Math.Pow(1 + 0.0001 * distance, -0.5);
                break;
        }
        
        return (Math.Max(1.0, sigmaY), Math.Max(1.0, sigmaZ));
    }

    public string DetermineRiskLevel(double concentration, Chemical chemical)
    {
        var threshold = chemical.ToxicityThreshold;
        
        if (concentration > threshold) return "HIGH";
        if (concentration > threshold * 0.1) return "MEDIUM";
        if (concentration > threshold * 0.01) return "LOW";
        return "MINIMAL";
    }

    private double CalculateGaussianConcentration(double releaseRate, double distance, double height, double windSpeed, string stabilityClass)
    {
        if (windSpeed <= 0) windSpeed = 0.1; // Avoid division by zero
        
        var (sigmaY, sigmaZ) = CalculateDispersionCoefficients(stabilityClass, distance, height);
        
        // Gaussian plume equation
        var concentration = (releaseRate / (2 * Math.PI * windSpeed * sigmaY * sigmaZ)) 
                          * Math.Exp(-Math.Pow(height, 2) / (2 * Math.Pow(sigmaZ, 2)));
        
        return Math.Max(0, concentration);
    }

    private double CalculateDistance(double lat1, double lon1, double lat2, double lon2)
    {
        // Haversine formula for distance calculation
        const double R = 6371000; // Earth radius in meters
        
        var lat1Rad = lat1 * Math.PI / 180;
        var lat2Rad = lat2 * Math.PI / 180;
        var deltaLat = (lat2 - lat1) * Math.PI / 180;
        var deltaLon = (lon2 - lon1) * Math.PI / 180;
        
        var a = Math.Sin(deltaLat / 2) * Math.Sin(deltaLat / 2) +
                Math.Cos(lat1Rad) * Math.Cos(lat2Rad) *
                Math.Sin(deltaLon / 2) * Math.Sin(deltaLon / 2);
        
        var c = 2 * Math.Atan2(Math.Sqrt(a), Math.Sqrt(1 - a));
        
        return R * c;
    }

    private string DetermineStabilityClass(WeatherData weather)
    {
        // Simplified stability class determination
        // Normally this would use solar radiation, cloud cover, etc.
        var windSpeed = weather.WindSpeed;
        
        if (windSpeed < 2) return "F"; // Very stable
        if (windSpeed < 3) return "E"; // Slightly stable
        if (windSpeed < 5) return "D"; // Neutral
        if (windSpeed < 6) return "C"; // Slightly unstable
        return "B"; // Moderately unstable
    }
    
    private double CalculateCrosswindDispersion(double distance, string stabilityClass)
    {
        // Pasquill-Gifford dispersion coefficients for sigma_y
        // Based on stability class and distance in meters
        double a, b;
        
        switch (stabilityClass.ToUpper())
        {
            case "A": // Very unstable
                a = 0.22; b = 0.0001;
                break;
            case "B": // Moderately unstable
                a = 0.16; b = 0.0001;
                break;
            case "C": // Slightly unstable
                a = 0.11; b = 0.0001;
                break;
            case "D": // Neutral (default)
                a = 0.08; b = 0.0001;
                break;
            case "E": // Slightly stable
                a = 0.06; b = 0.0001;
                break;
            case "F": // Moderately stable
                a = 0.04; b = 0.0001;
                break;
            default:
                a = 0.08; b = 0.0001; // Default to neutral
                break;
        }
        
        return a * distance * Math.Pow(1 + b * distance, -0.5);
    }
    
    private double CalculateVerticalDispersion(double distance, string stabilityClass)
    {
        // Pasquill-Gifford dispersion coefficients for sigma_z
        double a, b;
        
        switch (stabilityClass.ToUpper())
        {
            case "A": // Very unstable
                a = 0.20; b = 0.0;
                break;
            case "B": // Moderately unstable
                a = 0.12; b = 0.0;
                break;
            case "C": // Slightly unstable
                a = 0.08; b = 0.0002;
                break;
            case "D": // Neutral (default)
                a = 0.06; b = 0.0015;
                break;
            case "E": // Slightly stable
                a = 0.03; b = 0.0003;
                break;
            case "F": // Moderately stable
                a = 0.016; b = 0.0003;
                break;
            default:
                a = 0.06; b = 0.0015; // Default to neutral
                break;
        }
        
        return a * distance * Math.Pow(1 + b * distance, -0.5);
    }
    
    private double CalculateWindOrientedConcentration(
        double releaseRate, 
        double downwindDistance, 
        double crosswindOffset, 
        double effectiveHeight, 
        double windSpeed, 
        double sigmaY, 
        double sigmaZ)
    {
        if (windSpeed <= 0 || downwindDistance <= 0 || sigmaY <= 0 || sigmaZ <= 0)
            return 0;
        
        // Classic Gaussian plume equation with crosswind decay
        var crosswindTerm = Math.Exp(-0.5 * Math.Pow(crosswindOffset / sigmaY, 2));
        var verticalTerm = Math.Exp(-0.5 * Math.Pow(effectiveHeight / sigmaZ, 2));
        
        // Add ground reflection term
        var reflectionTerm = Math.Exp(-0.5 * Math.Pow((effectiveHeight) / sigmaZ, 2));
        
        var concentration = (releaseRate / (2 * Math.PI * windSpeed * sigmaY * sigmaZ)) *
                           crosswindTerm * (verticalTerm + reflectionTerm);
        
        return Math.Max(0, concentration * 1000000); // Convert to mg/m³
    }
}

File: chemical-dispersion-electron/src/main/services/DispersionModelingService.ts
/**
 * Dispersion Modeling Service for Electron main process
 * Implements Gaussian plume calculations and atmospheric dispersion modeling
 */

import { GaussianPlumeModel } from '../../shared/physics/GaussianPlumeModel';
import { 
  Release, 
  WeatherData, 
  Receptor, 
  DispersionResult, 
  DispersionCalculationResponse 
} from '../../shared/types';

export class DispersionModelingService {

  /**
   * Calculate Gaussian plume dispersion for specific receptors
   */
  public async calculateGaussianPlume(
    release: Release, 
    weather: WeatherData, 
    receptors: Receptor[]
  ): Promise<DispersionCalculationResponse> {
    const startTime = Date.now();
    
    try {
      const results: DispersionResult[] = [];
      
      // Calculate effective parameters
      const effectiveHeight = release.releaseHeight + this.calculatePlumeRise(release, weather);
      const effectiveRate = this.calculateEffectiveReleaseRate(release, weather);
      const stabilityClass = weather.stabilityClass || this.determineStabilityClass(weather);
      
      // Calculate concentration at each receptor
      for (const receptor of receptors) {
        const distance = this.calculateDistance(
          release.latitude, release.longitude,
          receptor.latitude, receptor.longitude
        );
        
        const bearing = this.calculateBearing(
          release.latitude, release.longitude,
          receptor.latitude, receptor.longitude
        );
        
        // Convert to downwind/crosswind coordinates
        const windRad = (weather.windDirection - 90) * Math.PI / 180;
        const receptorBearingRad = bearing * Math.PI / 180;
        const relativeBearing = receptorBearingRad - windRad;
        
        const x = distance * Math.cos(relativeBearing); // downwind
        const y = distance * Math.sin(relativeBearing); // crosswind
        
        if (x > 0) { // Only calculate for downwind receptors
          const concentration = this.calculateGaussianConcentration(
            x, y, effectiveHeight, effectiveRate,
            weather.windSpeed, stabilityClass
          );
          
          const result: DispersionResult = {
            id: 0,
            calculationTime: new Date(),
            latitude: receptor.latitude,
            longitude: receptor.longitude,
            height: 1.5, // Breathing height
            concentration: Math.max(0, concentration),
            concentrationUnit: "mg/m³",
            distanceFromSource: distance,
            directionFromSource: bearing,
            windSpeed: weather.windSpeed,
            windDirection: weather.windDirection,
            stabilityClass: stabilityClass,
            temperature: weather.temperature,
            exceedsToxicityThreshold: false, // Would need chemical toxicity data
            riskLevel: this.determineRiskLevel(concentration),
            modelUsed: "Gaussian Plume",
            createdAt: new Date(),
            releaseId: release.id,
            receptorId: receptor.id
          };
          
          results.push(result);
        }
      }
      
      const calculationTime = Date.now() - startTime;
      
      return {
        success: true,
        results,
        calculationTime
      };
      
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Calculation failed',
        calculationTime: Date.now() - startTime
      };
    }
  }

  /**
   * Calculate dispersion grid for visualization
   */
  public async calculateDispersionGrid(
    release: Release,
    weather: WeatherData,
    gridSize: number = 100,
    maxDistance: number = 10000
  ): Promise<DispersionCalculationResponse> {
    const startTime = Date.now();
    
    try {
      const results = GaussianPlumeModel.calculateDispersionGrid(
        release, weather, gridSize, maxDistance
      );
      
      const calculationTime = Date.now() - startTime;
      
      return {
        success: true,
        results,
        calculationTime
      };
      
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Grid calculation failed',
        calculationTime: Date.now() - startTime
      };
    }
  }

  /**
   * Calculate Gaussian concentration at a point
   */
  private calculateGaussianConcentration(
    x: number, y: number, h: number, Q: number,
    u: number, stabilityClass: string
  ): number {
    if (x <= 0 || u <= 0) return 0;
    
    // Get dispersion parameters
    const dispParams = this.getDispersionParameters(stabilityClass);
    const sigmaY = this.calculateSigmaY(x, dispParams);
    const sigmaZ = this.calculateSigmaZ(x, dispParams);
    
    if (sigmaY <= 0 || sigmaZ <= 0) return 0;
    
    // Gaussian plume equation
    const lateralTerm = Math.exp(-0.5 * Math.pow(y / sigmaY, 2));
    const verticalTerm = Math.exp(-0.5 * Math.pow(h / sigmaZ, 2)) + 
                        Math.exp(-0.5 * Math.pow(-h / sigmaZ, 2)); // Ground reflection
    
    const concentration = (Q / (2 * Math.PI * u * sigmaY * sigmaZ)) * lateralTerm * verticalTerm;
    
    return concentration * 1000; // Convert to mg/m³
  }

  /**
   * Get Pasquill-Gifford dispersion parameters
   */
  private getDispersionParameters(stabilityClass: string) {
    const params = {
      'A': { ay: 0.22, by: 0.0001, az: 0.20, bz: 0.0 },
      'B': { ay: 0.16, by: 0.0001, az: 0.12, bz: 0.0 },
      'C': { ay: 0.11, by: 0.0001, az: 0.08, bz: 0.0002 },
      'D': { ay: 0.08, by: 0.0001, az: 0.06, bz: 0.0015 },
      'E': { ay: 0.06, by: 0.0001, az: 0.03, bz: 0.0003 },
      'F': { ay: 0.04, by: 0.0001, az: 0.016, bz: 0.0003 }
    };
    
    return params[stabilityClass.toUpperCase() as keyof typeof params] || params['D'];
  }

  /**
   * Calculate sigma Y (horizontal dispersion coefficient)
   */
  private calculateSigmaY(x: number, dispParams: any): number {
    return dispParams.ay * x * Math.pow(1 + dispParams.by * x, -0.5);
  }

  /**
   * Calculate sigma Z (vertical dispersion coefficient)
   */
  private calculateSigmaZ(x: number, dispParams: any): number {
    return dispParams.az * x * Math.pow(1 + dispParams.bz * x, -0.5);
  }

  /**
   * Calculate effective release rate
   */
  private calculateEffectiveReleaseRate(release: Release, weather: WeatherData): number {
    // Basic calculation - could be enhanced with evaporation models, etc.
    if (release.releaseRate) {
      return release.releaseRate; // kg/s
    }
    
    if (release.totalMass) {
      // For instantaneous releases, assume 1-hour duration
      return release.totalMass / 3600; // kg/s
    }
    
    return 1.0; // Default 1 kg/s
  }

  /**
   * Calculate plume rise due to buoyancy and momentum
   */
  private calculatePlumeRise(release: Release, weather: WeatherData): number {
    // Simplified plume rise calculation
    // In reality would consider stack parameters, exit velocity, temperature, etc.
    
    if (release.initialTemperature && release.initialTemperature > weather.temperature + 10) {
      // Hot release - some buoyant rise
      const deltaT = release.initialTemperature - weather.temperature;
      return Math.min(50, deltaT * 0.5); // Max 50m rise
    }
    
    return 0; // Assume ground-level release
  }

  /**
   * Determine atmospheric stability class from weather data
   */
  private determineStabilityClass(weather: WeatherData): string {
    if (weather.stabilityClass) {
      return weather.stabilityClass;
    }
    
    // Simplified classification based on wind speed
    // In reality would consider solar radiation, cloud cover, time of day
    const windSpeed = weather.windSpeed;
    
    if (windSpeed < 2) return 'F'; // Very stable
    if (windSpeed < 3) return 'E'; // Stable
    if (windSpeed < 5) return 'D'; // Neutral
    if (windSpeed < 6) return 'C'; // Slightly unstable
    return 'B'; // Unstable
  }

  /**
   * Determine risk level based on concentration
   */
  private determineRiskLevel(concentration: number): string {
    if (concentration > 100) return 'CRITICAL';
    if (concentration > 10) return 'HIGH';
    if (concentration > 1) return 'MODERATE';
    if (concentration > 0.1) return 'LOW';
    return 'MINIMAL';
  }

  /**
   * Calculate distance between two lat/lng points
   */
  private calculateDistance(lat1: number, lng1: number, lat2: number, lng2: number): number {
    const R = 6371000; // Earth's radius in meters
    const dLat = (lat2 - lat1) * Math.PI / 180;
    const dLng = (lng2 - lng1) * Math.PI / 180;
    const a = Math.sin(dLat / 2) * Math.sin(dLat / 2) +
              Math.cos(lat1 * Math.PI / 180) * Math.cos(lat2 * Math.PI / 180) *
              Math.sin(dLng / 2) * Math.sin(dLng / 2);
    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
    return R * c;
  }

  /**
   * Calculate bearing from one point to another
   */
  private calculateBearing(lat1: number, lng1: number, lat2: number, lng2: number): number {
    const dLng = (lng2 - lng1) * Math.PI / 180;
    const lat1Rad = lat1 * Math.PI / 180;
    const lat2Rad = lat2 * Math.PI / 180;
    
    const y = Math.sin(dLng) * Math.cos(lat2Rad);
    const x = Math.cos(lat1Rad) * Math.sin(lat2Rad) -
              Math.sin(lat1Rad) * Math.cos(lat2Rad) * Math.cos(dLng);
    
    const bearing = Math.atan2(y, x) * 180 / Math.PI;
    return (bearing + 360) % 360;
  }
}